### PR TITLE
Provide mechanism to control whether or not to trim whitespace in @CsvSource and @CsvFileSource

### LINKED ISSUE #2420: Provide mechanism to control whether or not to trim whitespace in @CsvSource and @CsvFileSource
`CsvArgumentProvider` does not support configuration of the underlying `CsvParser` by univocity.

## Steps to reproduce

```java
    @ParameterizedTest
    @CsvSource(
        "        SW34NT,SW34NT   "
    )
    fun test(one: String, two: String) {
        assertThat(one).isEqualTo(two)
    }

```

This should **NOT** work, or at least the behavior should be configurable:

CsvParserSettings.ignoreTrailing/LeadingWhitespace = false

## Context

 - Used versions (Jupiter/Vintage/Platform): 5.7.0
 - Build Tool/IDE: IntelliJ

## Deliverables

- [ ] ...

### PR DESCRIPTION
**Pull request by somayaj**
## Overview

Added support for @CsvSource annotation.ignoreTrailingAndLeadingWhiteSpace property

Resolves #2420

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [X] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [X] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [X] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [X] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java:63]: Looks like you forgot to pass `ignoreTrailingAndLeadingWhitespace` to `CsvParserSettings`.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java
Comment (somayaj) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java:63]: Needed guidance there, do you want ignoreTrailingAndLeadingWhitespace as a property of CsvParserSettings. or does it translate to both of these booleans - setIgnoreTrailingWhitespaces() and setIgnoreLeadingWhitespaces()?

Comment (marcphilipp) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java:63]: Yes, the idea was to set both booleans based on the single flag from the annotation.

Comment (somayaj) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java:63]: Ok, fixed. The code is checked in. Thanks

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:703]: The test that uses this method is missing, isn't it?

**Review by marcphilipp — CHANGES_REQUESTED** [review]

**Comment by somayaj** [comment]
@marcphilipp will this be merged soon?

**Comment by marcphilipp** [comment]
I think the tests are still incomplete. Could you please respond to my comments above?

**Comment by somayaj** [comment]
i did add the tests with the last commit. If you can check that, thank you.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (marcphilipp) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:709]: These need corresponding test cases. See `executesWithCsvSource` in the enclosing class for an example.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/CsvArgumentsProviderTests.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/CsvArgumentsProviderTests.java:172]: Please also add test for when the new annotation attribute is set to `false`.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/CsvArgumentsProviderTests.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/CsvArgumentsProviderTests.java:197]: added, see the latest commit.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:717]: Yes, please see the commits. I accidently added the PR for another issue. have reverted it. Thanks for the review.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (juliette-derancourt) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:119]: This method is not necessary, as you can already set the `ignoreTrailingAndLeadingWhitespace` boolean with the `ignoreTrailingAndLeadingWhitespace()` method you introduced line 77.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (juliette-derancourt) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:26]: The introduction of this method makes the tests less readable, please inline `svSource().lines(lines).ignoreTrailingAndLeadingWhitespace(ignoreTrailingAndLeadingWhitespace).build()` in your test methods.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (juliette-derancourt) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:179]: Why do we need to modify this test?

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java
Comment (juliette-derancourt) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java:126]:
```suggestion
	 * <p>Defaults to {@code true}.
```

**Review by juliette-derancourt — CHANGES_REQUESTED** [review]
@somayaj Some more feedback on the PR :)

Also, shouldn't this behavior be added to `@CsvFileSource` as well?

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (juliette-derancourt) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:119]: See https://github.com/junit-team/junit5/pull/2491#discussion_r545350282

Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:119]: yes. but i need it for the "false" attribute value instead of using the setting in CsvSetting

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:26]: ok

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:179]: my bad, will fix. i think this was not intentional

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java
Comment (somayaj) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvSource.java]: done, please check the latest commit.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java]: done, please check the latest commit.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:26]: done, please check the latest commit.

**Comment by somayaj** [comment]
@marcphilipp @juliette-derancourt Hi, i've made the requested changes could you please review them? Thank you.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java:40]:
```suggestion
			annotation.maxCharsPerColumn(), annotation.ignoreTrailingAndLeadingWhitespace());
```

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:732]: Please also add corresponding tests that check these two methods.

**Review by marcphilipp — CHANGES_REQUESTED** [review]

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (somayaj) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:748]: done, sorry missed that.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java
Comment (somayaj) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvParserFactory.java]: done, sorry missed that.

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileSource.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileSource.java:150]: Please hard-wrap this line at 80 characters as per [our conventions](https://github.com/junit-team/junit5/blob/main/CONTRIBUTING.md#javadoc) (same for the other annotation).

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestIntegrationTests.java:19]: We don't use star imports.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/CsvFileArgumentsProviderTests.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/CsvFileArgumentsProviderTests.java:449]:
```suggestion

	@Test
```

File: junit-jupiter-params/src/test/resources/trailing-leading-spaces.csv
Comment (marcphilipp) [inline junit-jupiter-params/src/test/resources/trailing-leading-spaces.csv:1]: Please add some trailing and leading spaces here and adjust `CsvFileArgumentsProviderTests`.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (marcphilipp) [reply junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:26]: This isn't done, yet.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:176]: Please generally add new lines between methods.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java
Comment (marcphilipp) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/provider/MockCsvAnnotationBuilder.java:204]: You shouldn't have to add this method (similar to the comment regarding `@CsvSource` above).

**Review by marcphilipp — CHANGES_REQUESTED** [review]
@somayaj Please let us know if you have time to make the requested changes. Otherwise, we can take it from here.

**Comment by somayaj** [comment]
@marcphilipp @juliette-derancourt As suggested, please take it from here if it's easier.

**Comment by somayaj** [comment]
> @marcphilipp @juliette-derancourt As suggested, please take it from here if it's easier.

Thank you, please let me know if there's anything I can do here.

**Comment by juliette-derancourt** [comment]
@somayaj I'm on it :) 

Unfortunately, I just discovered a potential issue with the CSV parser we're using: https://github.com/uniVocity/univocity-parsers/issues/439

We'll have to wait for a response/fix before merging this.

**Comment by somayaj** [comment]
Sounds good, thanks for the heads up

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileSource.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileSource.java:159]: Should be `ignoreLeadingAndTrailingWhitespace`

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileSource.java
Comment (marcphilipp) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileSource.java:150]: Let's clarify what a quoted CSV column is, i.e. that it starts with a quotation character.

**Comment by juliette-derancourt** [comment]
This has been merged in 17bff6e734ba84553f3f49293c6e94d580202a44 and revised in 50480fd2f79a9891eaf11c3351320cf3e23d47a3. Documentation also updated in d64a4a575d9c41492fc2a9c2e11b6fa428865cdf.

@somayaj Thanks for the PR!

**Comment by somayaj** [comment]
Thank you! Juliette and Marc.

**Pull request closed without merging**
